src/plugins/http.js

```diff
@@ -7,66 +7,81 @@ const Tags = opentracing.Tags
 const FORMAT_HTTP_HEADERS = opentracing.FORMAT_HTTP_HEADERS
 
 function patch (http, tracer, config) {
-  this.wrap(http, 'request', request => makeRequestTrace(request))
-  this.wrap(http, 'get', get => makeRequestTrace(get))
+  this.wrap(http, 'request', makeRequestTrace)
+  this.wrap(http, 'get', makeGetTrace)
 
   function makeRequestTrace (request) {
     return function requestTrace (options, callback) {
-      const uri = extractUrl(options)
-      const method = (options.method || 'GET').toUpperCase()
+      const req = request.call(this, options, callback)
 
-      if (uri === `${tracer._url.href}/v0.3/traces`) {
-        return request.apply(this, [options, callback])
+      // The `https.request` function before Node v9 internally invokes `http.request`. When both modules have been
+      // loaded by the user (indirectly) and both modules have been patched by us, that would lead to us double-counting
+      // the request.
+      if (isTraced(req, options)) {
+        return req
+      }
+      // Here we mark the `req` object, as opposed to also marking the `options` object like we do in `get`. The problem
+      // is that if we’d mark `options` here, it would get lost while upstream `request` derives a clone from it and
+      // thus would never get to the recursive invocation of `request`.
+      markAsTraced(req)
+
+      const span = createSpan(options)
+      if (span) {
+        const headers = createHeaders(span, options)
+        for (const header in headers) {
+          req.setHeader(header, headers[header])
+        }
+        registerEventHandlers(req, span)
       }
+      return req
+    }
+  }
 
-      options = typeof options === 'string' ? url.parse(uri) : Object.assign({}, options)
+  // Regardless of what Node version we’re running on, `https.get` does not invoke our patched `https.request`, thus we
+  // need to patch it as well.
+  function makeGetTrace (get) {
+    return function getTrace (options, callback) {
+      options = typeof options === 'string' ? url.parse(options) : Object.assign({}, options)
       options.headers = options.headers || {}
 
-      const parentScope = tracer.scopeManager().active()
-      const parent = parentScope && parentScope.span()
-      const span = tracer.startSpan('http.request', {
-        childOf: parent,
-        tags: {
-          [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_CLIENT,
-          'service.name': getServiceName(tracer, config, options),
-          'resource.name': method,
-          'span.type': 'web',
-          'http.method': method,
-          'http.url': uri
-        }
-      })
+      // Prevent from double-counting in `request`.
+      markAsTraced(options)
```

Comment:
Why not just wait until `req` is returned and mark it there? (same as `request`)

Reply:
Because `get` calls `end` on the `req` object before returning it, which is too late to inject headers.

Reply:
Because `get()` internally calls `request()` we need to signal to our patched version of `request()` that this request is already being traced so it can skip it.

Reply:
Hmm it seems there is another problem with this, in that the span is created *after* calling `request()` and thus would exclude its execution time.

> if we’d mark `options` here, it would get lost while upstream `request` derives a clone from it

Is this done by the plugin or by the `https` module?

Also, wouldn't it be possible to simply detect the protocol in the `http` patch, so that if it's `https` it skips instrumentation (since it's the handler specifically for `http`). Then the`http` handler could simply use the `https` handler and wrap it with this check.

Reply:
> Hmm it seems there is another problem with this, in that the span is created _after_ calling `request()` and thus would exclude its execution time.

The span is created on line 50 _before_ invoking the actual implementation on lines 53 or 57.

>> if we’d mark `options` here, it would get lost while upstream `request` derives a clone from it
> Is this done by the plugin or by the `https` module?

The `https` module https://github.com/nodejs/node/blob/v8.11.4/lib/https.js#L226-L237

> Also, wouldn't it be possible to simply detect the protocol in the `http` patch, so that if it's `https` it skips instrumentation (since it's the handler specifically for `http`). Then the `http` handler could simply use the `https` handler and wrap it with this check.

Ah yes, I forgot to include this in my initial post, but my very first attempt basically did something along those lines, but it meant that:

1. When patching the `https` module we _also_ need to activate the `http` plugin to ensure that both would be patched. This worked fine, but led to friction when deciding what/how to unpatch.
2. Because things work differently on Node < and >= v9, we’d need to conditionally patch some modules in some cases and not others. In and of itself this wasn’t really a big issue, because other plugins have version requirements too, it just added to the growing feeling of this not being the cleanest solution. (I had patched `instrumenter.js` to use the node version when a module couldn’t be located in `node_modules`.)

There was another hard issue I ran into but can’t remember offhand any longer, but it was probably the straw that broke the camel’s back and made me decide to take another stab at it from a different approach that would be agnostic to version specific issues, which is this PR.

Reply:
> The span is created on line 50 before invoking the actual implementation on lines 53 or 57.

Sorry for the confusion I actually meant line 15

> The https module https://github.com/nodejs/node/blob/v8.11.4/lib/https.js#L226-L237

Since we always convert the options to an object, the property would be copied if defined as enumerable (or with a simple assignment). The reasoning behind using `Object.defineProperties` was to hide very large data that we might add from `console.log` such as a span. Also, in this case, the options object is a copy anyway so any value assigned here will never be seen by the user.

> When patching the https module we also need to activate the http plugin to ensure that both would be patched. This worked fine, but led to friction when deciding what/how to unpatch.

I don't see why this would be the case if each of them have their own individual wrappers.

> Because things work differently on Node < and >= v9, we’d need to conditionally patch some modules in some cases and not others.

I would argue that I'd prefer a clear line of code to do that then unclear code and a comment.

Reply:
> once I’m done with a patch for receiving sampling priority from the agent

@alloy Please definitely let me know when you start working on that to avoid duplicating work as we were planning on adding that in the next few days.

Reply:
@alloy Did you make any progress on sampling priority? Otherwise this was the next item on my priority list.

Reply:
A little bit. I’ll hit you up in Slack in a few minutes